contracts/extensions/votingReputation/VotingReputationStaking.sol

```diff
@@ -0,0 +1,244 @@
+/*
+  This file is part of The Colony Network.
+
+  The Colony Network is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.8.20;
+pragma experimental ABIEncoderV2;
+
+import "./../../colonyNetwork/IColonyNetwork.sol";
+import "./../../colony/ColonyRoles.sol";
+import "./../../common/ERC20Extended.sol";
+import "./../../tokenLocking/ITokenLocking.sol";
+import "./VotingReputationStorage.sol";
+
+contract VotingReputationStaking is VotingReputationStorage {
+
+  // Public
+
+  function stakeMotion(
+    uint256 _motionId,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _vote,
+    uint256 _amount,
+    bytes memory _key,
+    bytes memory _value,
+    uint256 _branchMask,
+    bytes32[] memory _siblings
+  )
+    public
+  {
+    Motion storage motion = motions[_motionId];
+
+    require(_vote <= 1, "voting-rep-bad-vote");
+    require(getMotionState(_motionId) == MotionState.Staking, "voting-rep-motion-not-staking");
+
+    uint256 requiredStake = getRequiredStake(_motionId);
+    uint256 amount = min(_amount, requiredStake - motion.stakes[_vote]);
+    require(amount > 0, "voting-rep-bad-amount");
+
+    uint256 stakerTotalAmount = stakes[_motionId][msgSender()][_vote] + amount;
+
+    require(
+      stakerTotalAmount <= checkReputation(motion.rootHash, motion.skillId, msgSender(), _key, _value, _branchMask, _siblings),
+      "voting-rep-insufficient-rep"
+    );
+    require(
+      stakerTotalAmount >= wmul(requiredStake, userMinStakeFraction) ||
+      (motion.stakes[_vote] + amount) == requiredStake, // To prevent a residual stake from being un-stakable
+      "voting-rep-insufficient-stake"
+    );
+
+    // Update the stake
+    motion.stakes[_vote] += amount;
+    stakes[_motionId][msgSender()][_vote] = stakerTotalAmount;
+
+    // Increment counter & extend claim delay if staking for an expenditure state change
+    if (
+      _vote == YAY &&
+      !motion.escalated &&
+      motion.stakes[YAY] == requiredStake &&
+      motion.altTarget == address(0x0) &&
+      ((_motionId > motionCountV10)
+        ? isExpenditureSig(motion.sig)
+        : isExpenditureSig(getActionSummary(motion.action, getTarget(motion.altTarget)).sig))
+    ) {
+      require(_motionId > motionCountV10, "voting-rep-invalid-motion");
```

Comment:
Okay, so we're saying that:

V9 motions, if they do not relate to expenditures, can complete as normal.

If they do relate to expenditures:
* In the voting stage: can proceed to completion (with old functionality)
* In the staking stage, if YAY not yet staked: cannot proceed further. Otherwise, can move to completion

I'm not wild that this allows people to stake on the `YAY` side, but just not to completion, nor that whether something can move out of staking is determined by which side has staked how much so far. Is a clearer demarcation possible?

Reply:
Sure, I can re-write that check to occur in all calls to `StakeMotion`. My motivation for doing it the way I did was to reduce the gas overhead of the check, since it would only be done in some subset of staking calls.

Reply:
Ultimately the goal here is to prevent v9 motions from locking the expenditure after we upgrade.

We have to also consider the scenario where the `YAY` side has already staked, and then the extension was upgraded. In that case, the expenditure will have already been locked, and so we should allow `NAY` staking to continue (otherwise upgrading the extension is a free win to the `YAY` side).

If the `YAY` side hasn't fully staked, then we should probably disallow staking entirely on the motion, on both sides.

Reply:
> If the YAY side hasn't fully staked, then we should probably disallow staking entirely on the motion, on both sides.

I'd be happier with this functionality (for motions that would lock expenditures)